### PR TITLE
[Enhancement] Support group by STRUCT/MAP

### PR DESCRIPTION
**Pull request by Seaven**
Fixes #
Support Group By Struct/Map Type

* support serilized method on struct/map
    * map serilized: will sort map-key first 
* support group by struct/map, use common template: serilized to string and group by it

Pay attention: if the key of map can't promise unique, the result of group by is unstable

BTW, JSON's implement(vpack) can't promise a stable hash/serilized result on binary data

like that:
```
MySQL test> select map2, st1 from sc2;
+-----------------------------+----------------------------------------------------------------------+
| map2                        | st1                                                                  |
+-----------------------------+----------------------------------------------------------------------+
| {2:[2,3,4],1:[1,2,3]}       | {"s1":1,"s2":[2,1,3],"s3":{2:2,1:1},"s4":{"e":3,"f":2}}              |
| {2:[2,null,4],1:[1,2,null]} | {"s1":1,"s2":[2,1,null],"s3":{2:2,1:1},"s4":{"e":null,"f":2}}        |
| {2:[3,2,4],1:[1,2,3]}       | {"s1":1,"s2":[2,1,3],"s3":{1:1,2:2},"s4":{"e":3,"f":2}}              |
| {2:[null,2,4],1:[1,2,null]} | {"s1":1,"s2":[2,1,null],"s3":{1:1,2:2},"s4":{"e":null,"f":2}}        |
| {2:[2,3,4],1:[1,2,3]}       | {"s1":1,"s2":[2,1,3],"s3":{2:2,1:1},"s4":{"e":3,"f":2}}              |
| {2:[2,null,4],1:[1,2,null]} | {"s1":1,"s2":[2,1,null],"s3":{2:2,1:1},"s4":{"e":null,"f":2}}        |
| {2:[2,3,4],1:[1,2,3]}       | {"s1":1,"s2":[2,1,3],"s3":{2:2,1:1},"s4":{"e":3,"f":2}}              |
| {2:[2,null,4],1:[1,2,null]} | {"s1":1,"s2":[2,1,null],"s3":{2:2,1:1},"s4":{"e":null,"f":2}}        |
| {2:[3,2,4],1:[1,2,3]}       | {"s1":1,"s2":[2,1,3],"s3":{1:1,2:2},"s4":{"e":1,"f":2}}              |
| {2:[null,2,4],1:[1,2,null]} | {"s1":1,"s2":[2,1,null],"s3":{1:1,2:2},"s4":{"e":1,"f":2}}           |
| {1:[1,2,3],2:[2,4,3]}       | {"s1":1,"s2":[1,2,3],"s3":{2:2,3:3,1:1},"s4":{"e":1,"f":2}}          |
| {1:[1,2,null],2:[2,4,null]} | {"s1":1,"s2":[1,2,null],"s3":{2:2,null:null,1:1},"s4":{"e":1,"f":2}} |
| {1:[1,2,3],2:[2,3,4]}       | {"s1":1,"s2":[2,1,3],"s3":{1:1,2:2},"s4":{"e":3,"f":2}}              |
| {1:[1,2,null],2:[2,null,4]} | {"s1":1,"s2":[2,1,null],"s3":{1:1,2:2},"s4":{"e":null,"f":2}}        |
| {1:[2,1,3],2:[2,3,4]}       | {"s1":1,"s2":[1,2,3],"s3":{2:2,1:1,3:3},"s4":{"e":1,"f":2}}          |
| {1:[2,1,null],2:[2,null,4]} | {"s1":1,"s2":[1,2,null],"s3":{2:2,1:1,null:null},"s4":{"e":1,"f":2}} |
| {1:[2,1,3],2:[2,4,3]}       | {"s1":1,"s2":[1,2,3],"s3":{2:2,1:1,3:3},"s4":{"e":1,"f":2}}          |
| {1:[2,1,null],2:[2,4,null]} | {"s1":1,"s2":[1,2,null],"s3":{2:2,1:1,null:null},"s4":{"e":1,"f":2}} |
| {1:[1,2,3],2:[2,4,3]}       | {"s1":1,"s2":[1,2,3],"s3":{2:2,3:3,1:1},"s4":{"e":1,"f":2}}          |
| {1:[1,2,null],2:[2,4,null]} | {"s1":1,"s2":[1,2,null],"s3":{2:2,null:null,1:1},"s4":{"e":1,"f":2}} |
+-----------------------------+----------------------------------------------------------------------+
20 rows in set
Time: 0.057s

MySQL test> select map2, st1,  count(1) from sc2 group by  map2, st1
+-----------------------------+----------------------------------------------------------------------+----------+
| map2                        | st1                                                                  | count(1) |
+-----------------------------+----------------------------------------------------------------------+----------+
| {1:[1,2,3],2:[2,3,4]}       | {"s1":1,"s2":[2,1,3],"s3":{1:1,2:2},"s4":{"e":3,"f":2}}              | 4        |
| {1:[1,2,null],2:[2,null,4]} | {"s1":1,"s2":[2,1,null],"s3":{1:1,2:2},"s4":{"e":null,"f":2}}        | 4        |
| {1:[1,2,3],2:[3,2,4]}       | {"s1":1,"s2":[2,1,3],"s3":{1:1,2:2},"s4":{"e":3,"f":2}}              | 1        |
| {1:[1,2,null],2:[null,2,4]} | {"s1":1,"s2":[2,1,null],"s3":{1:1,2:2},"s4":{"e":null,"f":2}}        | 1        |
| {1:[1,2,3],2:[3,2,4]}       | {"s1":1,"s2":[2,1,3],"s3":{1:1,2:2},"s4":{"e":1,"f":2}}              | 1        |
| {1:[1,2,null],2:[null,2,4]} | {"s1":1,"s2":[2,1,null],"s3":{1:1,2:2},"s4":{"e":1,"f":2}}           | 1        |
| {1:[1,2,3],2:[2,4,3]}       | {"s1":1,"s2":[1,2,3],"s3":{1:1,2:2,3:3},"s4":{"e":1,"f":2}}          | 2        |
| {1:[1,2,null],2:[2,4,null]} | {"s1":1,"s2":[1,2,null],"s3":{null:null,1:1,2:2},"s4":{"e":1,"f":2}} | 2        |
| {1:[2,1,3],2:[2,3,4]}       | {"s1":1,"s2":[1,2,3],"s3":{1:1,2:2,3:3},"s4":{"e":1,"f":2}}          | 1        |
| {1:[2,1,null],2:[2,null,4]} | {"s1":1,"s2":[1,2,null],"s3":{null:null,1:1,2:2},"s4":{"e":1,"f":2}} | 1        |
| {1:[2,1,3],2:[2,4,3]}       | {"s1":1,"s2":[1,2,3],"s3":{1:1,2:2,3:3},"s4":{"e":1,"f":2}}          | 1        |
| {1:[2,1,null],2:[2,4,null]} | {"s1":1,"s2":[1,2,null],"s3":{null:null,1:1,2:2},"s4":{"e":1,"f":2}} | 1        |
+-----------------------------+----------------------------------------------------------------------+----------+
12 rows in set
Time: 0.071s
```

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 5 / 6 (83.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|1|2|50.00%|[242]|
|:large_blue_circle:|com/starrocks/sql/analyzer/SelectAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|1|1|100.00%|[]|

File: be/src/column/map_column.cpp [hunk 1]
```diff
@@ -234,14 +235,29 @@ void MapColumn::remove_first_n_values(size_t count) {
...
+        for (uint32_t i = 0; i < map_size; i++) {
+            perm[i].index_in_chunk = offset + i;
+        }
+        Tie tie(map_size, 1);
+        std::pair<int, int> range{0, map_size};
+        auto st = sort_and_tie_column(false, _keys, SortDesc(true, true), perm, tie, range, false);
+        DCHECK(st.ok());
+    }
+
     for (size_t i = 0; i < map_size; ++i) {
```
Comment (satanson) [inline be/src/column/map_column.cpp:257]: For group-by queries, if its group-by-columns of json/struct types are not pruned or output, such as
```sql
select a from t group by json_column
```
can we just  cast struct/json column into string?

File: be/src/column/json_column.cpp [hunk 2]
```diff
@@ -85,4 +85,11 @@ uint32_t JsonColumn::serialize(size_t idx, uint8_t* pos) {
     return static_cast<uint32_t>(get_object(idx)->serialize(pos));
 }
 
+void JsonColumn::serialize_batch(uint8_t* dst, Buffer<uint32_t>& slice_sizes, size_t chunk_size,
+                                 uint32_t max_one_row_size) {
+    for (size_t i = 0; i < chunk_size; ++i) {
+        slice_sizes[i] += serialize(i, dst + i * max_one_row_size + slice_sizes[i]);
```
Comment (satanson) [inline be/src/column/json_column.cpp:91]: That each row occupies max_one_row_size bytes wastes a lot memory, so suggest to  sequeeze the unused bytes away.

File: be/src/column/map_column.cpp [hunk 3]
```diff
@@ -234,14 +235,29 @@ void MapColumn::remove_first_n_values(size_t count) {
...
+    SmallPermutation perm(map_size);
+    {
+        for (uint32_t i = 0; i < map_size; i++) {
+            perm[i].index_in_chunk = offset + i;
+        }
+        Tie tie(map_size, 1);
+        std::pair<int, int> range{0, map_size};
+        auto st = sort_and_tie_column(false, _keys, SortDesc(true, true), perm, tie, range, false);
+        DCHECK(st.ok());
+    }
```
Comment (satanson) [inline be/src/column/map_column.cpp:255]: why we should sort  map keys when serializing MapColumn, sorting is time-consuming, it should be delayed to the timepoint when it really required to sort.

Comment (Seaven) [reply be/src/column/map_column.cpp:255]: Serilized method only called when compute the aggregate's group by key or hash join, map keys need a stable order to match others, like 
```
select x from (
select map{1:1, 2:2} x 
union all 
select map{2:2, 1:1}
)
group by x
```

we need promise the group by map is a stable  data, so we need sort map keys

File: be/src/column/map_column.cpp (hunk 1 above)
Comment (Seaven) [reply be/src/column/map_column.cpp:257]: the current implement is what you say

File: be/src/column/json_column.cpp (hunk 2 above)
Comment (Seaven) [reply be/src/column/json_column.cpp:91]: yeah! you are right, it's will cost so many memory, but it's need modify the aggregate framework, I can modify it in other PR?

**Review by satanson — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 20 / 23 (86.96%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|1|2|50.00%|[242]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/PruneSubfieldRule.java|15|17|88.24%|[206, 207]|
|:large_blue_circle:|com/starrocks/sql/analyzer/SelectAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|1|1|100.00%|[]|

**Review by stdpain — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

File: be/src/column/json_column.cpp (hunk 2 above)
Comment (fzhedu) [reply be/src/column/json_column.cpp:91]: need add a TODO

File: be/src/column/struct_column.cpp [hunk 4]
```diff
@@ -290,7 +290,16 @@ size_t StructColumn::filter_range(const Filter& filter, size_t from, size_t to)
 }
 
 int StructColumn::compare_at(size_t left, size_t right, const Column& rhs, int nan_direction_hint) const {
-    DCHECK(false) << "Dont support it";
+    const auto& rhs_struct = down_cast<const StructColumn&>(rhs);
+    DCHECK(_fields.size() != rhs_struct._fields.size());
```
Comment (fzhedu) [inline be/src/column/struct_column.cpp:294]: _fields.size() <= rhs_struct._fields.size()?

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java [hunk 5]
```diff
@@ -747,9 +747,7 @@ public boolean canGroupBy() {
         if (isArrayType()) {
             return ((ArrayType) this).getItemType().canGroupBy();
         }
-        // TODO(mofei) support group by for JSON
-        return !isOnlyMetricType() && !isJsonType() && !isFunctionType() && !isBinaryType() && !isStructType() &&
-                !isMapType();
+        return !isOnlyMetricType() && !isJsonType() && !isFunctionType() && !isBinaryType();
```
Comment (fzhedu) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Type.java:750]: need check the nested type of map() or struct().

**Pull request merged**